src/main/java/org/openjdk/jextract/json/JSONObjects.java

```diff
@@ -0,0 +1,826 @@
+    }
+
+    // is this a Map/HashMap/LinkedHashMap?
+    private static boolean isMap(Class<?> cls) {
+        return cls == Map.class || cls == HashMap.class || cls == LinkedHashMap.class;
+    }
+
+    // is this String or CharSequence class?
+    private static boolean isStringOrCharSequence(Type type) {
+        return type == String.class || type == CharSequence.class;
+    }
+
+    // The ParamerizedType is expected to have only one type argument at most
+    private static Type getFirstTypeArgument(ParameterizedType pt) {
+        return getNthTypeArgument(pt, 0);
+    }
+
+    private static Type getSecondTypeArgument(ParameterizedType pt) {
+        return getNthTypeArgument(pt, 1);
+    }
+
+    private static Type getNthTypeArgument(ParameterizedType pt, int n) {
+        Type[] typeArgs = pt.getActualTypeArguments();
+        Type nthType = typeArgs.length > n ? typeArgs[n] : Object.class;
+        if (nthType instanceof WildcardType wt) {
+            Type[] upper = wt.getUpperBounds();
+            Type[] lower = wt.getLowerBounds();
+            if (upper.length > 0) {
+                nthType = upper[0];
+            } else if (lower.length > 0) {
+                nthType = lower[0];
+            } else {
+                nthType = Object.class;
+            }
+        }
+        return nthType;
+    }
+
+    // record helpers
+
+    // get the record's canonical constructor
+    private static Constructor<? extends Record> getRecordConstructor(Class<? extends Record> cls) {
+        Class<?>[] paramTypes = Arrays.stream(cls.getRecordComponents()).
+            map(RecordComponent::getType).
+            toArray(Class<?>[]::new);
+
+        try {
+            return cls.getConstructor(paramTypes);
+        } catch (ReflectiveOperationException roe) {
+            throw new RuntimeException("cannot find canonical constructor", roe);
+        }
+    }
+
+    // get the Record component value as a JSONValue
+    private static JSONValue getRecordProperty(Record record, RecordComponent rc) {
+        Object value = null;
+        try {
+            value = rc.getAccessor().invoke(record);
+        } catch (ReflectiveOperationException roe) {
+            throw new RuntimeException(roe);
+        }
+        return from(value);
+    }
+
+    // JSONValue to convenient Java objects conversion helpers
+
+    // create cannot convert exception
+    private static RuntimeException cannotConvert(JSONValue jv, Type targetType) {
+        return new RuntimeException("cannot convert " + jv + " to " + targetType.getTypeName());
+    }
+
+    // JSONObject converter helpers
+
+    // exists only localized SuppressWarnings
+    @SuppressWarnings("unchecked")
+    private static Record convertToRecord(JSONObject jo, Class<?> cls) {
+        assert cls.isRecord();
+        return toRecord(jo, (Class<? extends Record>)cls);
+    }
+
+    private static Map<?,?> convertToMap(JSONObject jo, Type keyType, Type valueType) {
```

Comment:
This seems overly general - isn't a JSON object always a map from String to something else?

Reply:
Keys can be anything convertible from String (like enum for eg).  Yes, there are not that many, but.. 

record MyRecord(Map<SymbolType, <valuetype>> ....)